### DISCUSSION TITLE
Support multimode keymapping like `vim.keymap.set()` and lazy.nvim

### DISCUSSION THREAD

**Discussion by Andrew15-5** [NvChad/NvChad#2489] [Ideas]
Is it possible to create a `{'n', 'v'}` type of format when creating keymaps?

```vim
vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, { buffer = true })
```

so maybe like this:

```lua
M.custom = {
  n = {},
  v = {},
  nv = {}, -- or vn
  -- or
  ["nv"] = {},
}
```

And individual mode keymaps will override what is written in the multimode keymaps.

**Comment by siduck** [comment]
@Andrew15-5  your syntax breaks nvchad mapping syntax btw. but you can achieve basic functionality  with lua  

![image](https://github.com/NvChad/NvChad/assets/59060246/2a6d4441-01ee-4d40-8e5e-b64c45b61dde)

**Reply by Andrew15-5** [reply]
> your syntax breaks nvchad mapping syntax btw

I know. This is why I opened an "idea" discussion and not a "Q&A". I thought that maybe such feature could be added as an alternative to `{'n', 'v'}` syntax.

The provided workaround at the bottom of the screen does achieve the same goal, but I need to first create the `t` element in `M.terminal` so that I can create something in it using `t["something"]`. And I also have to define a keymap inside of the `M.terminal.t` table and then reference it using `M.terminal.n["..."]`. I still would like to request the proposed feature. It won't break any existing config, but will add a similar (even more concise) syntax as the one used in aforementioned `vim.keymap.set()` and lazy.nvim (and probably in many other places/plugins). In case of the string `["nv"]` the assinged table of keymaps has to be applied to `n` and `v` mode. This IMO easier to implement than `nv` element (need to check all permutations or define a specific order of modes). If a keymap has to be set for many modes, then you can just do `["nxoti"]` or `inotx` (in alphabetical order).

**Reply by siduck** [reply]
@Andrew15-5 you can still use the vim.keymap syntax to add new mappings, but to override default nvchad mappings, how will you do that with vim.keymap.set? for that you have to use our own mapping syntax

**Reply by Andrew15-5** [reply]
> to override default nvchad mappings, how will you do that with vim.keymap.set?

You mean with `{"a", "b"}` syntax? Well, I already showed the syntax, here is a more elaborate version:

```lua
M.general = {
  ["nv"] = {
    ["<leader>lr"] = { "cmd", "Overrides deafult <leader>lr" },
  },
  -- instead of
  n = {
    ["<leader>lr"] = { "cmd", "Overrides deafult <leader>lr" },
  },
  v = {
    ["<leader>lr"] = { "cmd", "Overrides deafult <leader>lr" },
  },
}
```

and disable like this:

```lua
M.disabled = {
  ["nv"] = {
    ["<leader>lr"] = "",
  },
  -- or
  n = {
    ["<leader>lr"] = ""
  },
  v = {
    ["<leader>lr"] = "",
  },
}
```

So the `["nv"]` is just a wrapper for `n` and `v`, but literal `n` and `v` have an obvious priority (if both combined and literal versions are used).

**Reply by siduck** [reply]
nono,   i meant `vim.keymap.set({blabla}`, you can use this function, just put it in custom init.lua. Nvchad's syntax is just to override nvchad mappings or adding new mappings if u want to document them in nvcheatsheet

**Reply by Andrew15-5** [reply]
> adding new mappings if u want to document them in nvcheatsheet

That's the whole point. NvChad provides a central place where you can define shortcuts in a simple and concise way. Therefore, I don't want to deviate from it too much (some keymaps can't be easily set) and instead want to use the same `custom.mappings` file but with a slightly newer syntax to add the same keymap for different modes.

>  i meant `vim.keymap.set({blabla}`, you can use this function, just put it in custom init.lua

If you suggest to use `vim.keymap.set({blabla})` instead of NvChad's syntax (while using NvChad's base configs), then what's the point of `custom.mappings`?

**Reply by siduck** [reply]
you are right, i will have to modify some code of core.load_mappings which will take some time :/ , so in the mean time just use those work arounds

**Reply by Andrew15-5** [reply]
Great. I will create an issue then. Here: #2491.

**Reply by siduck** [reply]
@Andrew15-5 issue isnt needed for  this. i will decide whether to add this feature or not this weekend.

**Reply by Andrew15-5** [reply]
But if this is a feature request, then isn't that's why we use issue trackers? To track things like new features?

**Reply by siduck** [reply]
@Andrew15-5 this was feature was requested before too, i dont remember the issue id but it was closed probably cuz the codebase got a lil complex cuz of it, but i will look this weekend.

**Reply by Andrew15-5** [reply]
You are sneaky, @siduck! /s If you don't remember the closed issue id (or can't find it in general), then it is obviously easier to open a new one, but you just deleted it. I think you are afraid of new issues being visual present (in the https://github.com/NvChad/NvChad/issues). It reminded me of this meme format:

<details><summary>meme</summary>

![meme](https://github.com/NvChad/NvChad/assets/37143421/6b6a99ed-1729-47d3-8a3c-d599519a71ab)

</details>

**Reply by siduck** [reply]
@Andrew15-5 yes but I accept issues which are bug related or feature requests which will be 100% added. 

But i'm at a lil conflict over myself with this feature so there are 50% chances of it not being added so i dont consider an issue to be made for this :grin:

**Reply by Andrew15-5** [reply]
I can provide another 50% in a form of code contribution. :)

But not necessarily in the next few days (busy).
